sops/__init__.py

```diff
@@ -578,29 +567,58 @@ def encrypt_key_with_kms(key, entry):
     if 'arn' not in entry or entry['arn'] == "":
         print("KMS ARN not found, skipping entry", file=sys.stderr)
         return entry
+    kms = get_aws_session_for_entry(entry)
+    if kms is None:
+        panic("failed to initialize AWS KMS client for entry",
+              file=sys.stderr)
+    try:
+        kms_response = kms.encrypt(KeyId=entry['arn'], Plaintext=key)
+    except Exception as e:
+        print("failed to encrypt key using kms arn %s: %s, skipping it" %
+              (entry['arn'], e), file=sys.stderr)
+        return entry
+    entry['enc'] = b64encode(
+        kms_response['CiphertextBlob']).decode('utf-8')
+    entry['created_at'] = time.time()
+    return entry
+
+
+def get_aws_session_for_entry(entry):
+    """Return a boto3 session using a role if one exists in the entry"""
     # extract the region from the ARN
     # arn:aws:kms:{REGION}:...
     res = re.match('^arn:aws:kms:(.+):([0-9]+):key/(.+)$', entry['arn'])
     if res is None:
         print("Invalid ARN '%s' in entry" % entry['arn'], file=sys.stderr)
-        return entry
+        return None
     try:
         region = res.group(1)
     except:
         print("Unable to find region from ARN '%s' in entry" %
               entry['arn'], file=sys.stderr)
-        return entry
-    kms = boto3.client('kms', region_name=region)
+        return None
+    # if there are no role to assume, return the client directly
+    if not ('role' in entry):
+        return boto3.client('kms', region_name=region)
+    # otherwise, create a client using temporary tokens that assume the role
     try:
-        kms_response = kms.encrypt(KeyId=entry['arn'], Plaintext=key)
+        client = boto3.client('sts')
+        role = client.assume_role(RoleArn=entry['role'],
+                                  RoleSessionName='sops@'+gethostname())
     except Exception as e:
-        print("failed to encrypt key using kms arn %s: %s, skipping it" %
-              (entry['arn'], e), file=sys.stderr)
-        return entry
-    entry['enc'] = b64encode(
-        kms_response['CiphertextBlob']).decode('utf-8')
-    entry['created_at'] = time.time()
-    return entry
+        print("Unable to switch roles: %s" % e, file=sys.stderr)
```

Comment:
probably want to `return None` here to prevent it from entering the next try block